.github/workflows/pre-commit.yml

```diff
@@ -0,0 +1,24 @@
+# https://github.com/pre-commit/action
+
+name: "pre-commit"
+
+on:
+  push:
+    branches:
+      - master
+  pull_request: ~
```

Comment:
What is the intention of tilde (`~`) here? If it's intentional, we maybe need a comment.

Reply:
That means an empty list and it's part of markdown syntax. Not sure what comment is needed.

Reply:
Interesting! Let's drop it then, lone `pull_request:` works just fine.

Reply:
Turns out it's `None`, not an empty list. In Python:

```
In [1]: import yaml

In [2]: yaml.safe_load('foo: ~')
Out[2]: {'foo': None}

In [3]: yaml.safe_load('foo:')
Out[3]: {'foo': None}

In [4]: yaml.safe_load("foo: []")
Out[4]: {'foo': []}
```

And in the Yaml spec at https://yaml.org/spec/1.2-old/spec.html#id2805071 .

Reply:
@ulgens actually let's also drop the limit on `master` branch. We want CI everywhere. Plain…

```yaml
on:
  push:
  pull_request:
```

…will be perfect, if we add cron later.

Reply:
I'll check it again. The problem I was trying to solve there is, "on push, pull_request" config triggers all tests twice on the pull request with no visible benefit.

Reply:
@ulgens it only looks like running the same thing twice, it's actually two different things: One CI is testing the commit at the head of the pull request, the other is doing a temporary merge commit and tests the merged result. Often there's little difference, semantically one is testing the pull request in isolation and the other if merging will break the target branch. We want both.

Reply:
@ulgens for proof:

In pull request https://github.com/hartwork/jawanndenn/pull/411:
- The branch HEAD is at commit 4e9a6e7468e7564cf7018643cba2b14dbe064d24
- One CI run `Build and test using Docker` is testing that commit (see https://github.com/hartwork/jawanndenn/runs/5218188939?check_suite_focus=true#step:2:364)
- The other CI run `Build and test using Docker` is testing the merge result (see https://github.com/hartwork/jawanndenn/runs/5218189320?check_suite_focus=true#step:2:361)

Reply:
@ulgens maybe I was too quick here and my reply probably came across in a non-ideal way. I now understand the idea better and I see that it's a tradeoff, either too much CI or too little, either potentially waiting and more CO2 or more chance to miss some subtle breakage. Does it matter to you which approach we pick here? I'm more open to giving the current proposal a try now, if you want.

Reply:
@hartwork About push vs pull request, I don't think it matters on a repository that had no commits in the last 2.5 years. My intention with all of these commits is to recover the repo from dead and make it useful again, in the fastest way possible. Forking it was an option, but I didn't want to deal with submitting an orphan package on PyPI. That being said, top-notch code quality is always important but I don't think fixing an extra empty space on a commit or discussions about CI trigger config provides any extra value.

That being said, I tend to not having (seemingly) duplicate Cı checks but I'm okay with whatever you see fit.

Reply:
Hi @ulgens ,

> My intention with all of these commits is to recover the repo from dead and make it useful again, in the fastest way possible.

Personally, I want quality over speed. Why go as fast as possible? If my review is too picky, I can offer we flip this around i.e. you open bugs, I create PRs, and you get to take the review seat, instead. What do you think?

> Forking it was an option, but I didn't want to deal with submitting an orphan package on PyPI.

I'm not sure how orphan packages on PyPI work. Could you elaborate or share a link to what you refer to?

> That being said, top-notch code quality is always important but I don't think fixing an extra empty space on a commit or discussions about CI trigger config provides any extra value.

I think can agree to disagree here. Stray whitespace changes in commits are something that goes into Git history forever, and it's a time waster and annoying when inspecting past history, and it's cheap to fix and unnecessary in my  view. (I have a blog post on [my approach to code review](https://blog.hartwork.org/posts/my-approach-to-code-review/) if you're curious.)

> I tend to not have (seemingly) duplicate CI checks but I'm okay with whatever you see fit.

Okay, I would vote for the simpler but more covering version then, i.e.
```
on:
  push:
  pull_request:
```

Reply:
Fixed.